### PR TITLE
Support to backup and recover the specific graphs via vineyardctl

### PR DESCRIPTION
**Pull request by dashanji**
### <samp>🤖 Generated by Copilot at b435676</samp>

This pull request enhances the vineyard integration for GraphScope on Kubernetes. It enables backup and recovery of vineyard deployments using the `Session` class, and simplifies the code in the `Graph` class by removing a redundant import.

## Related issue number

Fixes parts of #2539 

Wait vineyard to release a new version.

### CODE REVIEW

File: python/graphscope/client/session.py [hunk 1]
```diff
@@ -1098,6 +1161,13 @@ def g(
         retain_oid=True,
         vertex_map="global",
     ):
+        if (
+            isinstance(incoming_data, vineyard.ObjectID)
+            and self._vineyard_object_mapping_table[str(incoming_data)] is not None
+        ):
+            incoming_data = vineyard.ObjectID(
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1168]: Log a message: f"Restoring graph from original graph {graph_id}"

File: python/graphscope/client/session.py [hunk 2]
```diff
@@ -1089,6 +1094,64 @@ def get_config(self):
         """Get configuration of the session."""
         return self._config_params
 
+    def backup_vineyard_depployment(self, graphIDs: str, path: str, pvc_name: str):
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1097]: The `graphIds` seems is a list, why there is a str? If you accept both str and list, please address this in the comment.
Also, prefer detailed comments for these two methods, if you would like people to understand and using these.

File: python/graphscope/client/session.py [hunk 3]
```diff
@@ -1089,6 +1094,64 @@ def get_config(self):
...
+        if self._cluster_type != types_pb2.K8S:
+            return
+        # check if session exist
+        if self.session_id is None:
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1101]: You could not both to consider such case like the session is invalid or something.

File: python/graphscope/client/session.py [hunk 4]
```diff
@@ -1089,6 +1094,64 @@ def get_config(self):
         """Get configuration of the session."""
         return self._config_params
 
+    def backup_vineyard_depployment(self, graphIDs: str, path: str, pvc_name: str):
+        if self._cluster_type != types_pb2.K8S:
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1098]: raise an error... do not fail silently.

File: python/graphscope/client/session.py [hunk 5]
```diff
@@ -1089,6 +1094,64 @@ def get_config(self):
...
+            return
+        # check if session exist
+        if self.session_id is None:
+            return
+        vineyard_deployment_name = self._config_params["k8s_vineyard_deployment"]
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1103]: Do you need to check if there is a vineyard deployment? The session may be launched with vineyard as a sidecar container.

**Review by siyuan0322 — APPROVED** [review]

File: python/graphscope/client/session.py (hunk 5 above)
Comment (dashanji) [reply python/graphscope/client/session.py:1103]: Thanks reminding. I will add a check here.

File: python/graphscope/client/session.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dashanji) [reply python/graphscope/client/session.py:1097]: Yes, `to_pvc` looks good to me, I will change it to `store_graph_to_pvc` and `restore_graph_from_pvc`.

File: python/graphscope/client/session.py [hunk 6]
```diff
@@ -1098,6 +1096,147 @@ def get_config(self):
...
+        try:
+            api_client = resolve_api_client(self._config_params["k8s_client_config"])
+        except kube_config.ConfigException as e:
+            raise RuntimeError(
+                "Kubernetes environment not found, you may want to"
+                ' launch session locally with param cluster_type="hosts"'
+            ) from e
+        return api_client
+
+    def _check_pvc_exist(self, pvc_name, pvc_namespace):
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1109]: **exists**

File: python/graphscope/client/session.py [hunk 7]
```diff
@@ -1098,6 +1096,147 @@ def get_config(self):
...
+                       - list: list of vineyard.ObjectID
+            path: The path in the pv to which the pvc is bound.
+            pvc_name: The name of the PVC.
+
+        Raises:
+            RuntimeError: If the cluster type is not Kubernetes.
+        """
+        if self._cluster_type != types_pb2.K8S:
+            raise RuntimeError("Only support kubernetes cluster")
+        if not isinstance(graphIDs, list):
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1167]: Do not raise error.
```python
if not isintance(objects, list):
  objects = [objects]
```

File: python/graphscope/client/session.py [hunk 8]
```diff
@@ -1098,6 +1096,147 @@ def get_config(self):
...
+            raise RuntimeError("graphIDs should be a list of vineyard.ObjectID")
+        objectids = ",".join(repr(id) for id in graphIDs)
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1169]: objectids -> object_ids

File: python/graphscope/client/session.py [hunk 9]
```diff
@@ -1098,6 +1096,147 @@ def get_config(self):
...
+            backup_name="vineyard-backup-" + random_string(6),
+            vineyard_deployment_name=vineyard_deployment_name,
+            vineyard_deployment_namespace=namespace,
+            namespace=namespace,
+            path=path,
+            objectids=objectids,
+            pvc_name=pvc_name,
+        )
+
+    def restore_graphs_from_pvc(self, path: str, pvc_name: str):
```
Comment (siyuan0322) [inline python/graphscope/client/session.py:1186]: restore_from_pvc

File: python/graphscope/client/session.py [hunk 10]
```diff
@@ -1098,6 +1096,147 @@ def get_config(self):
...
+            )
+        except kube_client.rest.ApiException as e:
+            if e.status == 404:
+                raise RuntimeError(
+                    f"Vineyard deployment {vineyard_deployment_name} not found in namespace {vineyard_deployment_namespace}"
+                ) from e
+            else:
+                raise e
+
+    def store_graphs_to_pvc(self, graphIDs, path: str, pvc_name: str):
```
[the remark this answers is not in the document]
Comment (dashanji) [reply python/graphscope/client/session.py:1145]: Make sense.

**Review by sighingnow — APPROVED** [review]

**Comment by sighingnow** [comment]
@dashanji Could you please try to rebase against the latest master? Thanks!

**Comment by dashanji** [comment]
> @dashanji Could you please try to rebase against the latest master? Thanks!

Okay.

**Pull request merged**